src/controller/java/CHIPDeviceController-JNI.cpp

```diff
@@ -958,19 +958,54 @@ JNI_METHOD(void, releaseOperationalDevicePointer)(JNIEnv * env, jobject self, jl
     }
 }
 
-JNI_METHOD(void, shutdownSubscriptions)(JNIEnv * env, jobject self, jlong handle, jlong devicePtr)
+JNI_METHOD(jint, getFabricIndex)(JNIEnv * env, jobject self, jlong handle)
 {
     chip::DeviceLayer::StackLock lock;
 
-    DeviceProxy * device = reinterpret_cast<DeviceProxy *>(devicePtr);
+    AndroidDeviceControllerWrapper * wrapper = AndroidDeviceControllerWrapper::FromJNIHandle(handle);
+
+    return wrapper->Controller()->GetFabricIndex();
+}
 
-    //
-    // We should move away from this model of shutting down subscriptions in this manner and instead,
-    // have Java own the ReadClient objects directly and manage their lifetimes.
-    //
-    // #13163 tracks this issue.
-    //
-    device->ShutdownSubscriptions();
+JNI_METHOD(void, shutdownSubscriptions)
+(JNIEnv * env, jobject self, jobject handle, jobject fabricIndex, jobject peerNodeId, jobject subscriptionId)
+{
+    // ShutdownSubscription should be called from the Matter thread with the lock held
+    assertChipStackLockedByCurrentThread();
```

Comment:
How do we make sure this JNI is always called from the Matter thread?

Reply:
unfortunately there is no way to guarantee it is called from matter thread, but we could assert in advance, matter is based upon single thread assumption, it is application's responsibility to post the work into matter thread.

Reply:
Ok, I see you have removed assertChipStackLockedByCurrentThread();